awx/main/tasks.py

```diff
@@ -952,184 +954,247 @@ def run(self, pk, **kwargs):
                 raise RuntimeError('AWX_PROOT_BASE_PATH=%s does not exist' % settings.AWX_PROOT_BASE_PATH)
 
             # store a record of the venv used at runtime
-            if hasattr(instance, 'custom_virtualenv'):
-                self.update_model(pk, custom_virtualenv=getattr(instance, 'ansible_virtualenv_path', settings.ANSIBLE_VENV_PATH))
-
-            # Fetch ansible version once here to support version-dependent features.
-            kwargs['ansible_version'] = get_ansible_version()
-            kwargs['private_data_dir'] = self.build_private_data_dir(instance, **kwargs)
+            if hasattr(self.instance, 'custom_virtualenv'):
+                self.update_model(pk, custom_virtualenv=getattr(self.instance, 'ansible_virtualenv_path', settings.ANSIBLE_VENV_PATH))
+            private_data_dir = self.build_private_data_dir(self.instance)
 
             # Fetch "cached" fact data from prior runs and put on the disk
             # where ansible expects to find it
-            if getattr(instance, 'use_fact_cache', False):
-                instance.start_job_fact_cache(
-                    os.path.join(kwargs['private_data_dir']),
-                    kwargs.setdefault('fact_modification_times', {})
+            if getattr(self.instance, 'use_fact_cache', False):
+                self.instance.start_job_fact_cache(
+                    os.path.join(private_data_dir, 'artifacts', str(self.instance.id), 'fact_cache'),
+                    fact_modification_times,
                 )
 
             # May have to serialize the value
-            kwargs['private_data_files'] = self.build_private_data_files(instance, **kwargs)
-            kwargs['passwords'] = self.build_passwords(instance, **kwargs)
-            kwargs['proot_show_paths'] = self.proot_show_paths
-            if getattr(instance, 'ansible_virtualenv_path', settings.ANSIBLE_VENV_PATH) != settings.ANSIBLE_VENV_PATH:
-                kwargs['proot_custom_virtualenv'] = instance.ansible_virtualenv_path
-            args = self.build_args(instance, **kwargs)
-            safe_args = self.build_safe_args(instance, **kwargs)
-            output_replacements = self.build_output_replacements(instance, **kwargs)
-            cwd = self.build_cwd(instance, **kwargs)
-            env = self.build_env(instance, **kwargs)
+            private_data_files = self.build_private_data_files(self.instance, private_data_dir)
+            passwords = self.build_passwords(self.instance, kwargs)
+            proot_custom_virtualenv = None
+            if getattr(self.instance, 'ansible_virtualenv_path', settings.ANSIBLE_VENV_PATH) != settings.ANSIBLE_VENV_PATH:
+                proot_custom_virtualenv = self.instance.ansible_virtualenv_path
+            self.build_extra_vars_file(self.instance, private_data_dir, passwords)
+            args = self.build_args(self.instance, private_data_dir, passwords)
+            # TODO: output_replacements hurts my head right now
```

Comment:
Do we need this to sanitize scm passwords in URLs?  I've always been a little fuzzy on why this code existed here in the first place.

Reply:
Would like to see this answered/addressed pre-merge or shortly thereafter.

Reply:
Good question.